File: backend/app/schemas/project.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import model_validator

from app.enums.allocation_type import AssignmentAllocationType, PhaseAllocationType
from app.enums.billing_type import BillingType
from app.enums.project_status import PhaseStatus, ProjectStatus
from app.schemas.base import JsonModel
from app.schemas.role_requirement import RoleRequirementCreate, RoleRequirementResponse


class PhaseAssignmentInput(JsonModel):
    """Inline assignment to create alongside a phase."""

    employee_id: UUID
    allocation_type: AssignmentAllocationType
    allocated_hours: int | None = None
    allocation_percentage: int | None = None
    hourly_rate_override: Decimal | None = None
    start_date: date | None = None
    end_date: date | None = None


class PhaseCreate(JsonModel):
    name: str
    start_date: date
    end_date: date | None = None
    allocation_type: PhaseAllocationType
    required_hours: int | None = None
    required_headcount: int | None = None
    budget: Decimal | None = None
    phase_order: int = 0
    status: PhaseStatus = PhaseStatus.PLANNING
    role_requirements: list[RoleRequirementCreate] = []
    new_assignments: list[PhaseAssignmentInput] = []


class PhaseUpdate(JsonModel):
    name: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    allocation_type: PhaseAllocationType | None = None
    required_hours: int | None = None
    required_headcount: int | None = None
    budget: Decimal | None = None
    phase_order: int | None = None
    status: PhaseStatus | None = None


class PhaseResponse(JsonModel):
    id: UUID
    project_id: UUID
    name: str
    start_date: date
    end_date: date | None
    allocation_type: PhaseAllocationType
    required_hours: int | None
    required_headcount: int | None
    budget: Decimal | None
    phase_order: int
    status: PhaseStatus
    role_requirements: list[RoleRequirementResponse] = []
    created_at: datetime
    updated_at: datetime


class ProjectCreate(JsonModel):
    name: str
    customer: str
    description: str | None = None
    salesforce_link: str | None = None
    start_date: date
    end_date: date | None = None
    billing_type: BillingType = BillingType.TIME_AND_MATERIALS
    fixed_price_amount: Decimal | None = None
    funding_source_id: UUID | None = None
    phases: list[PhaseCreate] = []

    @model_validator(mode="after")
    def validate_fixed_price(self) -> "ProjectCreate":
        if self.billing_type == BillingType.FIXED_PRICE:
            if self.fixed_price_amount is None or self.fixed_price_amount <= 0:
                msg = "fixed_price_amount is required and must be > 0 for fixed-price projects"
                raise ValueError(msg)
        return self


class ProjectUpdate(JsonModel):
    name: str | None = None
    customer: str | None = None
    description: str | None = None
    salesforce_link: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    status: ProjectStatus | None = None
    billing_type: BillingType | None = None
    fixed_price_amount: Decimal | None = None
    funding_source_id: UUID | None = None
    phases: list[PhaseCreate] | None = None


class ProjectResponse(JsonModel):
    id: UUID
    name: str
    customer: str
    description: str | None
    salesforce_link: str | None
    start_date: date
    end_date: date | None
    status: ProjectStatus
    billing_type: BillingType
    fixed_price_amount: Decimal | None
    funding_source_id: UUID | None
    funding_source_name: str | None
    created_at: datetime
    updated_at: datetime


class ProjectWithPhasesResponse(ProjectResponse):
    phases: list[PhaseResponse] = []

File: client/src/types/project.ts
export const BillingType = {
  FIXED_PRICE: "fixed_price",
  TIME_AND_MATERIALS: "time_and_materials",
} as const;

export type BillingType = (typeof BillingType)[keyof typeof BillingType];

export const ProjectStatus = {
  PLANNING: "planning",
  ACTIVE: "active",
  COMPLETED: "completed",
  WAITING_ASSIGNMENT: "waiting_assignment",
} as const;

export type ProjectStatus = (typeof ProjectStatus)[keyof typeof ProjectStatus];

export const PhaseStatus = {
  PLANNING: "planning",
  ACTIVE: "active",
  COMPLETED: "completed",
  WAITING_ASSIGNMENT: "waiting_assignment",
} as const;

export type PhaseStatus = (typeof PhaseStatus)[keyof typeof PhaseStatus];

export const PhaseAllocationType = {
  HOURS: "hours",
  HEADCOUNT: "headcount",
  HOURS_AND_HEADCOUNT: "hours_and_headcount",
} as const;

export type PhaseAllocationType =
  (typeof PhaseAllocationType)[keyof typeof PhaseAllocationType];

export interface RoleRequirement {
  id: string;
  projectId: string;
  phaseId: string | null;
  roleId: string;
  roleName: string;
  allocationPercentage: number;
  count: number;
  createdAt: string;
  updatedAt: string;
}

export interface RoleRequirementCreate {
  roleId: string;
  allocationPercentage: number;
  count: number;
}

export interface Phase {
  id: string;
  projectId: string;
  name: string;
  startDate: string;
  endDate: string | null;
  allocationType: PhaseAllocationType;
  requiredHours: number | null;
  requiredHeadcount: number | null;
  budget: number | null;
  phaseOrder: number;
  status: PhaseStatus;
  roleRequirements: RoleRequirement[];
  createdAt: string;
  updatedAt: string;
}

export interface PhaseAssignmentInput {
  employeeId: string;
  allocationType: "percentage" | "hours";
  allocatedHours?: number;
  allocationPercentage?: number;
  hourlyRateOverride?: number;
  startDate?: string;
  endDate?: string;
}

export interface PhaseCreate {
  name: string;
  startDate: string;
  endDate?: string;
  allocationType: PhaseAllocationType;
  requiredHours?: number;
  requiredHeadcount?: number;
  budget?: number;
  phaseOrder: number;
  status?: PhaseStatus;
  roleRequirements?: RoleRequirementCreate[];
  newAssignments?: PhaseAssignmentInput[];
}

export interface Project {
  id: string;
  name: string;
  customer: string;
  description: string | null;
  salesforceLink: string | null;
  startDate: string;
  endDate: string | null;
  status: string;
  billingType: BillingType;
  fixedPriceAmount: number | null;
  fundingSourceId: string | null;
  fundingSourceName: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface ProjectWithPhases extends Project {
  phases: Phase[];
}

export interface ProjectCreate {
  name: string;
  customer: string;
  description?: string;
  salesforceLink?: string;
  startDate: string;
  endDate?: string;
  billingType?: BillingType;
  fixedPriceAmount?: number;
  fundingSourceId?: string;
  phases: PhaseCreate[];
}

export interface ProjectUpdate {
  name?: string;
  customer?: string;
  description?: string;
  salesforceLink?: string;
  startDate?: string;
  endDate?: string;
  status?: string;
  billingType?: BillingType;
  fixedPriceAmount?: number;
  fundingSourceId?: string;
  phases?: PhaseCreate[];
}
